### ISSUE TITLE
Segmentation Fault at loss.backward

### ISSUE THREAD

**Issue by Crysitna** [pytorch/xla#901]
I'm using the following code
```
def train_loop_fn(model, loader, device, context): 
    model.train()
    for i, data in loader:
          outp = model(data)
          loss = loss_fn(outp)
          loss.backward()           <------------ segmentation fault
          xm.optimizer_step(optimizer)

model = Model(xxx)
devices = xm.get_xla_supported_devices(max_devices=8)
c= dp.DataParallel(model, device_ids=devices)

model_parallel(train_loop_fn, datagen)
```
Using Prebuild VM Image and tested on both `pytorch-0.1` and `pytorch-nightly` environment. 
The loss function is self-implemented, yet both model and loss function has been tested before on gpu. any suggestion?

**Comment by ibeltagy** [comment]
It could be a bug in the implementation of one of the operations. I would start with a simple model and loss function that work, then step by step add more operations to the model until I find the problematic one.

**Comment by Crysitna** [comment]
oh I seems found the reason, it's the loss function causing the error: 
```
def hinge_loss_wrong(pos, neg):
    device, dtype = pos.device, pos.dtype
    B = pos.size(1)

    ans = torch.zeros(batch_size).type(dtype).to(device)
    ans[:] = torch.clamp(1 - positive_t1[:] + negative_t1[:], min=0)

    return torch.sum(ans) / B
    

def hinge_loss_correct(pos, neg):
    device, dtype = pos.device, pos.dtype
    B = pos.size(1)
    
    ans = torch.clamp(1 - positive_t1[:] + negative_t1[:], min=0)
    return torch.sum(ans) / B
```
yet dont quite understand why the new defined tensor would cause a segmentation fault here?

**Comment by dlibenzi** [comment]
Any chance to get a repro?
Or a GDB backtrace (`bt`) of the faulting thread?

**Comment by Crysitna** [comment]
for repro, say just replace the loss_fn in mnist example with
```
def loss_fn(outp, targ):
  l = torch.zeros_like(outp).to(outp.device)
  l[:] = torch.log(outp) * targ[:,None]
  # return -(outp.log() * targ[:,None]).mean()   # WORKS
  return l.mean()                                               # SEGMENTATION FAULT
```

**Comment by dlibenzi** [comment]
I have verified that this does not crash with current code.
